File: JLHorizontalSelector/ControlView.m
//
//  ControlView.m
//  JLHorizontalSelector
//
//  Created by 廖宗綸 on 2015/6/12.
//  Copyright (c) 2015年 TSUNG-LUN LIAO. All rights reserved.
//

#import "ControlView.h"

@interface ControlView ()

@property (nonatomic, strong) UIImageView *theImageView;
@property (nonatomic, strong) id<ControlViewDelegate> delegate;

@end

@implementation ControlView

+ (ControlView*)creatControlViewWithImageName:(NSString*)imageName withDelegate:(id)delegate{
    
    ControlView *controlView = [[ControlView alloc] initWithImageName:imageName withDelegate:delegate];
    
    return controlView;
}

- (id)initWithImageName:(NSString*)imageName withDelegate:(id)delegate{
    
    self = [super initWithFrame:CGRectMake(0, 0, 0, 0)];
    
    _theImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:imageName]];
    
    self.frame = _theImageView.frame;
    
    [self addImageView];
   
    
    
    self.delegate = delegate;
//    self.backgroundColor = [UIColor grayColor];
    
    return self;
}



- (void)addImageView{
    
    [self addSubview:_theImageView];
    
    [_theImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    NSLayoutConstraint *constraint_Top = [NSLayoutConstraint constraintWithItem:_theImageView
                                                                      attribute:NSLayoutAttributeTop
                                                                      relatedBy:NSLayoutRelationEqual
                                                                         toItem:self
                                                                      attribute:NSLayoutAttributeTop
                                                                     multiplier:1.0
                                                                       constant:0];
    
    NSLayoutConstraint *constraint_Buttom = [NSLayoutConstraint constraintWithItem:_theImageView
                                                                         attribute:NSLayoutAttributeBottom
                                                                         relatedBy:NSLayoutRelationEqual
                                                                            toItem:self
                                                                         attribute:NSLayoutAttributeBottom
                                                                        multiplier:1.0
                                                                          constant:0];
    
    NSLayoutConstraint *constraint_Left = [NSLayoutConstraint constraintWithItem:_theImageView
                                                                       attribute:NSLayoutAttributeLeft
                                                                       relatedBy:NSLayoutRelationEqual
                                                                          toItem:self
                                                                       attribute:NSLayoutAttributeLeft
                                                                      multiplier:1.0
                                                                        constant:0];
    
    NSLayoutConstraint *constraint_Right = [NSLayoutConstraint constraintWithItem:_theImageView
                                                                        attribute:NSLayoutAttributeRight
                                                                        relatedBy:NSLayoutRelationEqual
                                                                           toItem:self
                                                                        attribute:NSLayoutAttributeRight
                                                                       multiplier:1.0
                                                                         constant:0];
    
    [self addConstraints:@[constraint_Top,constraint_Buttom,constraint_Left,constraint_Right]];
    
    _theImageView.layer.cornerRadius = (_theImageView.frame.size.width+_theImageView.frame.size.height)/4;
    _theImageView.clipsToBounds = YES;
    
    
}

#pragma mark - Touch Event
- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event{
    
    NSLog(@"Start Selector");
    [self.delegate controlView:self didBeginTouch:touches];
    
}

- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event{
    
    NSLog(@"Move...");
    [self.delegate controlView:self didMoveTouch:touches];
    
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event{
    
    NSLog(@"Select Done");
    [self.delegate controlView:self didStopTouch:touches];
    
}





@end

File: Swift/JLHorizontalSelector_Swift/JLHorizontalSelector_Swift/ControlView.swift
//
//  ControlView.swift
//  JLHorizontalSelector_Swift
//
//  Created by TSUNG-LUN LIAO on 2015/6/18.
//  Copyright (c) 2015年 TSUNG-LUN LIAO. All rights reserved.
//


import UIKit
//import Cocoa


protocol ControlViewDelegate{
    
    
    func controlView(controlView:ControlView?, didBeginTouch touches:NSSet?)
    func controlView(controlView:ControlView?, didMoveTouch  touches:NSSet?)
    func controlView(controlView:ControlView?, didStopTouch  touches:NSSet?)

}

class ControlView: UIView {
    
// MARK: Private Variable
    private var deleagte        :ControlViewDelegate?
    private var theImageView    :UIImageView!

    init(imageName:NSString, delegate:AnyObject){
        
        super.init(frame: CGRectMake(CGFloat(0), CGFloat(0), CGFloat(0), CGFloat(0)))
        
        self.deleagte = delegate as? ControlViewDelegate
        
        self.theImageView = UIImageView(image: UIImage(named: imageName as String))
        self.frame = self.theImageView.frame
        
        //Add ImageView
        self.addImageView()
        
        
        
    }

    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }


    
    func addImageView(){
        
        self.addSubview(self.theImageView)
        self.setTranslatesAutoresizingMaskIntoConstraints(false)
//        self.translatesAutoresizingMaskIntoConstraints = false
        
        let constraints_Top:NSLayoutConstraint = NSLayoutConstraint(
            item: self.theImageView,
            attribute: NSLayoutAttribute.Top,
            relatedBy: NSLayoutRelation.Equal,
            toItem: self,
            attribute: NSLayoutAttribute.Top,
            multiplier: 1.0,
            constant: 0)
        
        let constraints_Bottom:NSLayoutConstraint = NSLayoutConstraint(
            item: self.theImageView,
            attribute: NSLayoutAttribute.Bottom,
            relatedBy: NSLayoutRelation.Equal,
            toItem: self,
            attribute: NSLayoutAttribute.Bottom,
            multiplier: 1.0,
            constant: 0)
        
        let constraints_Left:NSLayoutConstraint = NSLayoutConstraint(
            item: self.theImageView,
            attribute: NSLayoutAttribute.Left,
            relatedBy: NSLayoutRelation.Equal,
            toItem: self,
            attribute: NSLayoutAttribute.Left,
            multiplier: 1.0,
            constant: 0)
        
        let constraints_Right:NSLayoutConstraint = NSLayoutConstraint(
            item: self.theImageView,
            attribute: NSLayoutAttribute.Right,
            relatedBy: NSLayoutRelation.Equal,
            toItem: self,
            attribute: NSLayoutAttribute.Right,
            multiplier: 1.0,
            constant: 0)
        
        self.addConstraints([constraints_Top,constraints_Bottom,constraints_Left,constraints_Right])
        
        self.theImageView.layer.cornerRadius = (self.theImageView.frame.size.width+self.theImageView.frame.size.height)/4
        self.theImageView.clipsToBounds = true
        
    }

// MARK: - Touch Event
    
    override func touchesBegan(touches: Set<NSObject>, withEvent event: UIEvent?) {
        
        NSLog("Start Selector")
        self.deleagte!.controlView(self, didBeginTouch: touches)
        
    }
    
    override func touchesMoved(touches: Set<NSObject>, withEvent event: UIEvent?) {
        
        NSLog("Move...")
        self.deleagte!.controlView(self, didMoveTouch: touches)
        
    }
    
    override func touchesEnded(touches: Set<NSObject>, withEvent event: UIEvent?) {
        
        NSLog("Select Done")
        self.deleagte!.controlView(self, didStopTouch: touches)
        
    }


}
